JIT compiler + bugfixes

eh-steve opened the pull request:
This is mostly a rebased commit history of #62, plus the addition of a new `jit` package to allow compiling and loading Go files/packages/text, and recursively resolving their dependencies.

This allows users to build and load arbitrary Go files with any number of imports automatically.

You can simply do:

```go

loadable, err = jit.BuildGoFiles(jit.BuildConfig{}, "./path/to/file1.go", "/path/to/file2.go")
// or
loadable, err = jit.BuildGoPackage(jit.BuildConfig{}, "./path/to/package")
// or
loadable, err = jit.BuildGoText(jit.BuildConfig{}, `
package mypackage

import "encoding/json"

func MyFunc(input []byte) (interface{}, error) {
	var output interface{}
	err := json.Unmarshal(input, &output)
	return output, err
}

`)

if err != nil {
    panic(err)
}

module, symbols, err = loadable.Load()

if err != nil {
    panic(err)
}

MyFunc := symbols["MyFunc"].(func([]byte) (interface{}, error))
result, err := MyFunc([]byte(`{"k":"v"}`))
```

The `jit` package and tests are early on in the commit history to demonstrate the purpose of each subsequent commit in fixing specific tests.


So far, testing has only been on `linux/amd64`, `linux/arm64`, `darwin/amd64`, `darwin/arm64` (Apple M1 silicon), and `windows/amd64` . Once the last few `goloader` bugs are ironed out, I will try to backport these changes for earlier Go versions where possible.

Versions:


| **OS/Arch**        | amd64/+CGo         | arm64/+CGo         | amd64/-CGo         | arm64/-CGo         |
|--------------------|--------------------|--------------------|--------------------|--------------------|
| Linux/go-1.19.4    | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Darwin/go-1.19.4   | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:            | :heavy_check_mark:               |
| Windows/go-1.19.4  | :heavy_check_mark:               | :interrobang:      | :heavy_check_mark:                | :interrobang:      |
| Linux/go-1.18.8    | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| Darwin/go-1.18.8   | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:            | :heavy_check_mark:                |
| Windows/go-1.18.8  | :x:                | :interrobang:      | :heavy_check_mark:                | :interrobang:      |
| Linux/go-1.17.13   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.17.13  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.17.13 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.16.15   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.16.15  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.16.15 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.15.15   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.15.15  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.15.15 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.14.15   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.14.15  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.14.15 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.13.15   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.13.15  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.13.15 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.12.17   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.12.17  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.12.17 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.11.13   | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.11.13  | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.11.13 | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.10.8    | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.10.8   | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.10.8  | :x:                | :x:                | :x:                | :x:                |
| Linux/go-1.9.7     | :x:                | :x:                | :x:                | :x:                |
| Darwin/go-1.9.7    | :x:                | :x:                | :x:                | :x:                |
| Windows/go-1.9.7   | :x:                | :x:                | :x:                | :x:                |

Commit:
Add support for deduplicating type descriptors by relocating code to point at main module's types

Commit:
Fix implementation of replacing local package names with import paths (fixes cases where import path doesn't match the local package name)

Commit:
Add support for checking unresolved symbols and reporting back all loaded packages

Commit:
RODATA is where type descriptors will be - we should load these too

Commit:
Don't add to the package list if the only thing included is the init func

Commit:
Improve error

Commit:
Split previously pclntab into separate tables (as done in runtime) to assist in debugging + fix cutab layout to correctly offset when creating multi-object, multi-package linkers - previously stack traces had incorrect files due to incorrect cutab + filetab

Commit:
Fix bug with deferred returns not being added due to pointer equality instead of string equality

Commit:
Add support for ASM function wrappers (ABIInternal -> ABI0) and return Munmap error

Commit:
Add support for loading native archive entries (e.g. CGo DNS resolution)

Commit:
Add patch of first module's itabs to relocate unreachable methods to the new module's code address

Commit:
Add JIT compiler and test cases

pkujhd committed:
delete useless code. typemap already is filled in buildModule

pkujhd commented:
@eh-steve , thanks.
I will try to Recurrence bugs and fix.

By the way, goloader does not test with cgo and asm.

pkujhd committed:
Update README.md

eh-steve commented:
> By the way, goloader does not test with cgo and asm.
I appreciate that - my main motivation was getting common packages like `net/http` and `math` to build successfully (which directly or indirectly use CGo/asm), and my fixes do seem to work (as shown in the tests).

Interestingly, after rebasing this onto the new Apple silicon commits in master, the http test now fails (where previously it was passing). I kept this working branch as a reference:
https://github.com/eh-steve/goloader/tree/jit-before-apple-silicon-rebase

pkujhd commented:
@eh-steve
I try use your jit branch to run go test on jit dir, every test output errors like "jit_test.go:42: still have 17 unresolved external symbols despite building and linking dependencies..."

Could you give me an simple case for TestStackSplit and TestJitPanicRecoveryStackTrace this issue mentioned?
Or tell me how to run your test.

Thx

pkujhd committed:
delete useless code

eh-steve commented:
What version of Go are you using, how do you run the tests, and which symbols remain unresolved? (I'm using 1.19.3 and it's working fine)

I've made certain core packages forbidden for re-inclusion in the JIT compiler to prevent duplication of certain atomic or runtime packages, are the missing symbols from these forbidden packages?

If so, they just need to be included in the test binary by declaring some global variables which use them.

Commit:
Fix occasional error "go.info.runtime.itabTable: relocation target go.info.*github.com/pkujhd/goloader.itabTableType not defined" caused by linker not generating dwarf info for goloader.itabTableType correctly. Also fixes for go.info.runtime.itabLock

eh-steve commented:
Edit: 
Fixed this now

> Some tests are still failing randomly (e.g. `TestStackSplit` and `TestJitPanicRecoveryStackTrace`) due to some outstanding bugs in goloader which still need to be fixed.
> 
> Specifically:
> Things which trigger stack copies/stack splits while a dynamically loaded function is running can cause an incomplete stack unwinding (`fatal error: traceback did not unwind completely`) , possibly because the `g`'s stack pointer was incorrectly incremented twice somewhere, so the unwinding doesn't ever reach the stack start. Unclear whether this is a TLS relocation issue or something else causing the `sp` to be incremented by the same frame size twice when calling dynamic code?
> 
> ### Example:
> ```
> === RUN   TestJitPanicRecoveryStackTrace/BuildGoFiles
> runtime: g808: frame.sp=0xc004befd60 top=0xc004beffe0
> 	stack=[0xc004bec000-0xc004bf0000] n=4 max=2147483647
> fatal error: traceback did not unwind completely
> 
> runtime stack:
> runtime.throw({0x849c62?, 0xc00024e000?})
> 	/usr/local/go/src/runtime/panic.go:1047 +0x5d fp=0x7fb3f6ffca88 sp=0x7fb3f6ffca58 pc=0x439e7d
> runtime.gentraceback(0x7fb40c71400b?, 0xb94c30?, 0xc00003ef00?, 0x7fb3f6ffce60?, 0x0, 0x0, 0x7fffffff, 0x7fb3f6ffce48, 0x1?, 0x0)
> 	/usr/local/go/src/runtime/traceback.go:561 +0x1385 fp=0x7fb3f6ffcdf8 sp=0x7fb3f6ffca88 pc=0x460d85
> runtime.addOneOpenDeferFrame.func1()
> 	/usr/local/go/src/runtime/panic.go:645 +0x6b fp=0x7fb3f6ffce70 sp=0x7fb3f6ffcdf8 pc=0x43900b
> runtime.systemstack()
> 	/usr/local/go/src/runtime/asm_amd64.s:492 +0x49 fp=0x7fb3f6ffce78 sp=0x7fb3f6ffce70 pc=0x46c269
> 
> goroutine 808 [running]:
> runtime.systemstack_switch()
> 	/usr/local/go/src/runtime/asm_amd64.s:459 fp=0xc004befb20 sp=0xc004befb18 pc=0x46c200
> runtime.addOneOpenDeferFrame(0x41c3e6?, 0xbac840?, 0xc004befbe8?)
> 	/usr/local/go/src/runtime/panic.go:644 +0x69 fp=0xc004befb60 sp=0xc004befb20 pc=0x438f49
> panic({0x7a6480, 0x8dd510})
> 	/usr/local/go/src/runtime/panic.go:844 +0x112 fp=0xc004befc20 sp=0xc004befb60 pc=0x439792
> command-line-arguments.(*SomeType).callSite4(0xc00027bd20?, {{0x7a6480, 0x8dd510}, 0x0, {0x0, 0x0, 0x0}, {0x0, 0x0}, 0x0})
> 	/tmp/file1_1093370123/file4.go:16 +0x25 fp=0xc004befc40 sp=0xc004befc20 pc=0x417352ad
> command-line-arguments.(*SomeType).callSite3(0x4e2e73746e656d75?, {{0x7a6480, 0x8dd510}, 0x0, {0x0, 0x0, 0x0}, {0x0, 0x0}, 0x0})
> 	/tmp/file1_1093370123/file3.go:13 +0x50 fp=0xc004befca0 sp=0xc004befc40 pc=0x41735268
> command-line-arguments.(*SomeType).callSite2(0xc000028034?, {{0x7a6480, 0x8dd510}, 0x0, {0x0, 0x0, 0x0}, {0x0, 0x0}, 0x0})
> 	/tmp/file1_1093370123/file2.go:11 +0x50 fp=0xc004befd00 sp=0xc004befca0 pc=0x417351f8
> command-line-arguments.(*SomeType).callSite1(0x0?, {{0x7a6480, 0x8dd510}, 0x0, {0x0, 0x0, 0x0}, {0x0, 0x0}, 0x0})
> 	/tmp/file1_1093370123/file1.go:7 +0x50 fp=0xc004befd60 sp=0xc004befd00 pc=0x41735188
> created by testing.(*T).Run
> 	/usr/local/go/src/testing/testing.go:1493 +0x35f
> 
> ```

eh-steve commented:
@pkujhd All tests are passing on my machine now, I think it's ready for review

Commit:
Add proper TLS resolution for externally linked binaries

Commit:
Bugfix type registration and improve deduplication (not all pointer types were stepped into, and not all types were registered correctly)

pkujhd commented:
> Edit: Fixed this now
> 
> > Some tests are still failing randomly (e.g. `TestStackSplit` and `TestJitPanicRecoveryStackTrace`) due to some outstanding bugs in goloader which still need to be fixed.

how to fixed it? I don't find the code change for goloader.

pkujhd commented:
@eh-steve 

For commit https://github.com/pkujhd/goloader/pull/66/commits/31f8f2bd0af2c793c20cae1e47c9507343b52b25, the pclntable does not split for adapter golang version < 1.16

 For commit [3cd5a08](https://github.com/pkujhd/goloader/pull/66/commits/3cd5a08266674fa56cecae4569b6e391106b0a8d)， plugin use deduplicate function to deal it, but in goloader it will be lead to far address relocate error on R_ADDROFF & R_METHODOFF.  By modifying read-only memory. used first module type descriptor. it could avoid it, but I worried it will be forbidden on MacOSX. I think maybe the best way are use deduplicate function to deal it  and copy type descriptor with far address relocated from the first module into data segment. (it maybe means a lots of workload, because a far address type descriptor maybe is an interface)

For commit [1f39a49](https://github.com/pkujhd/goloader/pull/66/commits/1f39a494d8d5c0a61a6d0d1888112ef63043aec0), could you give me a test case for recurrence it.

By the way, 
I cannot merge pr because the current version cannot pass all tests. I will wait for the newer commit.

pkujhd commented:
> I could fix this by forcing the symbols to be included by importing more things in `jit_test.go` if you want? In the meantime it should work with `-c`

it is worked for `go test -c`

eh-steve commented:
> how to fixed it? I don't find the code change for goloader.

Ah sorry, I rebased and force pushed my branch so it got hidden in the other commits. 

The problem was here in `initLinker()`. The runtime uses zero value offsets of `funcnametab` and `pctab` entries to mean something special (zero offset funcname means empty string, and zero offset pctab entry means an external non-Go function which we can't find stack information for).

The logic inside `runtime/traceback.go` (which is used in lots of places for stack copies etc) will be broken if a function has a zero offset for its `nameoff` or a zero offset for it's `pcsp` entry.

To fix it, I added an 8 byte padding to both of those tables so that valid entries never have an offset of 0.
```go
func initLinker() *Linker {
	linker := &Linker{
		// Pad these tabs out so offsets don't start at 0, which is often used in runtime as a special value for "missing"
		// e.g. runtime/traceback.go and runtime/symtab.go both contain checks like:
		// if f.pcsp == 0 ...
		// and
		// if f.nameoff == 0
		funcnametab:  make([]byte, PtrSize),
		pctab:        make([]byte, PtrSize),
```

Commit:
Make mprotect readonly by default, and add MprotectMakeExecutable

Commit:
Add reflectlite - a stripped back version of reflect package (with edits to allow setting unexported fields/types), and add support for preserving state across modules via new ConvertTypesAcrossModules() function

Commit:
Make string container configurable per linker/module, and add option to use heap allocated strings to be managed via GC

Commit:
Add go binary config to jit

Commit:
Make TestJITHttpGet more reliable by ensuring idle connection handling goroutines are closed before unloading

Commit:
Prevent infinite recursion in reflectlite

Commit:
Bugfix JIT - incorrect package names preventing init funcs from running

Commit:
Expose parsed files in LoadableUnit, and unload data on error during Load()

Commit:
Fix firstmodule itab patching to be safely reversible, and assorted other bugfixes

Commit:
Don't try to open executable as ELF on darwin

Commit:
Make symbol order in linker deterministic to help reproduce linker bugs (previously map iteration order would produce random variations, with rare but occasional errors)

Commit:
Add test for open coded defers - currently fails if linker symbol order is randomised?

Commit:
WIP - mmap manager + implement faster version of vmmap via Cgo

Commit:
Fix clobbered arm64 ADRP relocations during type deduplication

Commit:
Add lookup for CGo symbols from //go:cgo_import_dynamic pragma on Darwin

Commit:
Add binary self-patch for darwin to allow read only Mach-O segments to become writeable via mprotect

Commit:
Make libc work for all architectures, and support without CGo

Commit:
Use the non-default tls.Config in http test to prevent contaminating it with cross-module data

Commit:
Add build tag to disable darwin patch and skip CGo test on darwin + arm64

Commit:
Implement darwin vmmap fallback when Cgo is disabled

Commit:
Fixes for darwin/arm64

Commit:
Fix incorrect execve args for darwin

Commit:
Expose buildEnv in jit tests

Commit:
Add problem test for linux/arm64

Commit:
Word align data symbols for arm64

Commit:
unused import

Commit:
Copy + Sign + Replace patched executable mach-o file for darwin/arm64 because kernel caches codesigning info by inode

Commit:
Fix type conversion to support method closures (kinda)

Commit:
Make test fail

Commit:
Make inter-module function conversion even more dangerous

Commit:
Oops

Commit:
Fix ADRP idempotency

Commit:
arm64 BL instruction clobbers X30 (link register) - use B instead

Commit:
Avoid any writes to executable code on darwin/arm64

Commit:
Adjust placeholder native object code copy to avoid invalid slice on darwin/arm64

Commit:
Add JIT example to README.md

pkujhd commented:
@eh-steve  thanks for your contribute.

About per-linker string container with its own lifecycle. it should not be released when the module unload, because maybe some string already set into the map/array container in firstmodule,  when next code access it, the process will be panic.

Commit:
Don't free errors provided by libdl, and do register CGo symbols on other platforms

eh-steve commented:
> @eh-steve thanks for your contribute.
> 
> About per-linker string container with its own lifecycle. it should not be released when the module unload, because maybe some string already set into the map/array container in firstmodule, when next code access it, the process will be panic.

Yeah that’s why there’s a separate method to unload the string container, and also why I added the heap string option (which is much nicer to use)

Commit:
Actually implement ELF relocations for amd64 and arm64

Commit:
Stabilise some tests and commit relocation/deduplication debug writer

Commit:
Get x86 Mach-O relocations working

Commit:
arm64 mach-o relocs

eh-steve commented:
> I've now added basic native ELF relocs for arm64 and amd64, and that seems to make the CGo test work, I'll try to add the macho equivalent shortly.

Mach-O now added too (tested on both x86 and apple M1 silicon).

Commit:
Improve UnresolvedExternalSymbols by describing what uses the symbol

Commit:
Fix deduplication debug logger

Commit:
Experiment - try adding a relocation epilogue to every function which might need it, and use that epilogue to store any extra JIT instructions needed during far address relocation. Also patch all pcvalue data to ensure that if runtime unwinds the stack or pre-empts the function while it's executing these extra JIT instructions, pcvalue lookups don't fail
(previously this could cause fatal "runtime: invalid pc-encoded table" errors if we were unlucky enough to be caught by a pre-emption or a profiling signal whilst executing inside these instructions, since the runtime didn't know they existed since they appear to be outside of any pcvalue tables and findfuncbuckets)

Commit:
Put R_CALL back

Commit:
Try word aligning arm64 call epilogues

eh-steve commented:
@pkujhd You might also be interested in this bugfix:
https://github.com/pkujhd/goloader/pull/66/commits/b42e18025d406e7b279f1f11adf19ca43900fb05

Previously it was rare, but possible to be pre-empted by the scheduler (or a signal handler) while executing JIT instructions appended to the end of the code segment due to a far address relocation. In these cases, the runtime could throw a fatal error `runtime: invalid pc-encoded table` because the PC addresses don't correspond to any known functions based on `findfuncbucket`s, and they would be missing pcvalues for things like PCSP, PCLine, PCFile, and `PCDATA_UnsafePoint`. This wasn't safe.

I switched to pre-emptively growing the size of any function text which had a potentially overflowing relocation, and reserved an "epilogue" for each relocation to use if needed. Then I patched all the pcvalue tables in order to reflect the correct value for this epilogue PC segment (based on the pcvalue at the relocation site). This seems to have resolved these pre-emption bugs.

Commit:
Backup PCREL instructions 2 bytes before the actual reloc

Commit:
Adjust alignment of functions and epilogues

Commit:
Try making darwin arm64 mmapper more fussy about address hints

Commit:
Reduce number of syscalls when patching itabs

Commit:
Bugfix arm64 asm - use X17 (IP1 - intra-procedure-call corruptible register) instead of X9 for R_CALLARM64 relocs, since Go only uses them for the stack prologue and epilogue calculations, and we should already be clear of that by the time we hit an R_CALLARM64, so we should be able to safely use them for far jumps

Commit:
Simplify ADRP reloc code thanks to bidi linkless branch

Commit:
Bugfix linux/arm64 - need to flush instruction cache after relocation and before using code. Added asm along the lines of gcc's __aarch64_sync_cache_range() builtin

Commit:
Missing brackets

Commit:
Experiment - rewrite PCREL relocation implementation to use shorter asm, and don't use worst case epilogue size, instead compute the actual size needed for each reloc type (results in ~33% smaller code segment for x86, with more code fitting into cache as a result)

Commit:
Don't align code on 8 byte boundaries unnecessarily for x86 (save some more space)

Commit:
Add missing NOFRAME to mmap_linux_arm64.s read_CTR_ELO_Register()

Commit:
Fix accidental itab lock copy

Commit:
Bump go.mod to 1.9 for MacOS

Commit:
Add SkipTypeDeduplicationForPackages linker option and make JIT copy follow symlinks

Commit:
fix copy symlink resolution

Commit:
Add more granular control over whether to use duplicate symbols or firstmodule copy - also rerun inittasks for packages using duplicate symbols

Commit:
Improve type conversion for interface{} nested values

Commit:
WIP - mmap manager and mprotect for Windows

eh-steve commented:
@pkujhd FYI today I started to port some of the missing mprotect and mmap functionality for Windows in this branch https://github.com/eh-steve/goloader/commit/079d833490608d4ce1883081609b84c712d73874

It’s not finished and so doesn’t work properly yet, but just a heads up in case you were interested

Commit:
Add missing case for converting eface

Commit:
Simplify by using VirtualQueryEx across the entire address space, and switch page size to SystemInfo.AllocationGranularity (64k instead of 4k page size)

eh-steve commented:
> @pkujhd FYI today I started to port some of the missing mprotect and mmap functionality for Windows in this branch [eh-steve@079d833](https://github.com/eh-steve/goloader/commit/079d833490608d4ce1883081609b84c712d73874)
> 
> It’s not finished and so doesn’t work properly yet, but just a heads up in case you were interested

@pkujhd I finished this off today so most of the JIT tests pass on Windows now (and it solves previous problematic method relocations overflowing due to the distance being >32 bit), but proper CGo support on Windows is still missing, and I'm unlikely to invest time in adding it (because 🪟 💩).

Also, the `TestJitHttpGet` test currently fails on Windows with `all goroutines are asleep - deadlock` which I haven't investigated properly.

I'm also unlikely to invest much more time in backporting changes to older versions of Go... Is this something you feel is critical for the project?

eh-steve commented:
> > I'm also unlikely to invest much more time in backporting changes to older versions of Go... Is this something you feel is critical for the project?
> 
> 
> 
> because some related online projects still use golang 1.16 even golang 1.11.

Ok, what do you think about tagging the 0.0.17 release as the last version supporting Go < 1.18, and directing people to use that version if they need older versions of Go?

pkujhd committed:
fix #68 adapter golang 1.20

(cherry picked from commit 1fc223086769ff5c4ad989aa0949e3d4a17798fd)

pkujhd commented:
@eh-steve 
I will take the time to merge this branch into the trunk and try to adapt to the lower versions of Golang,
I think we should separate a new repos from the part of code（package jit) that provides compilation obj file, which is of a tool.
I will not consider ending the support of the lower version of golang, because there are still many bugs in the current version. Consider maintaining a new repos when golang 2.0

eh-steve commented:
I wish you would not reimplement parts of my commits manually as it makes it very difficult to rebase and resolve conflicts, unless you plan to repeat the  implementation of the entire PR by hand?

If this is the case, I might just hard fork and maintain a separate repo...

I don't mind splitting the JIT package into a separate repo, but the tests there have exposed maybe ~20 or so bugs in goloader and provides a nice test base when porting for new Go versions or architectures.

pkujhd commented:
@eh-steve I suggest you have a hard fork if you want make it work on golang version two major versions and  with cgo. 

I don't have enough to read all golang's linker，so I worry more complex mechanism to lead more unknown question.

online project use it on linux/amd64 and mmap on <4G address, avoid rewrite machine code as far as possible

In addition, thanks for fixing various bugs.

eh-steve commented:
Also I think you force pushed stuff to master?

pkujhd commented:
> Also I think you force pushed stuff to master?

I will block master branch before merge it， and try to merge it and split package. but i will take some time.

eh-steve commented:
There's still some broken things in Go 1.20 which I'm fixing - `inlinedCall` struct has changed size which isn't in master yet - I might be force pushing the last commit on this branch a few times while I fix the remaining tests for 1.20.

I meant that you had force pushed onto master which means any branches off master would have conflicts, and that it's generally not a good idea to force push onto master without careful coordination with other contributors.

I think you and I have a different understanding of "merge"...

Commit:
Edit readme

Commit:
Hard fork due to difficulty in collaboration

Commit:
Import paths varint encoded

Commit:
Expose module mapped addresses to allow users to know whether a type or string is inside a JIT module (and therefore unsafe to unload)

Commit:
Add support for R_ARM_PCREL_LDST relocs in Go 1.20 (porting https://github.com/pkujhd/goloader/commit/35cf5436c6ac065dce9abb1861de91fa1e0ff046)

pkujhd commented:
@eh-steve ，
 I suggest you have a hard fork, because When I merge you jit branch and try to adapter lower golang version， I find it very difficult. 
I will merge your bug fixed into the trunk. 
Thank you for your contribution.

Commit:
Register libresolv_* Cgo symbols on darwin

eh-steve commented:
Yeah I hard forked in https://github.com/pkujhd/goloader/pull/66/commits/e0dbc397477f5f6955a87a344b8539d40f188175

eh-steve commented:
> I will merge your bug fixed into the trunk.

I'd be curious how you plan to address the incorrect cuFiles and cuOffsets after the pclntab separation causing broken stack traces

Commit:
Bake in syscall.syscall9 on darwin on Go 1.20

eh-steve committed:
Merge branch 'master' into jit

eh-steve committed:
Merge pull request #1 from eh-steve/jit

JIT compiler

pkujhd commented:
> I'd be curious how you plan to address the incorrect cuFiles and cuOffsets after the pclntab separation causing broken stack traces

In the current version, there is a problem with the filling of ftab. If the loader does not use cuOffsets in the lower version of golang, the loader need to modify the PCFile. In the higher version of golang, cuOffsets exist. the loader can fill cuOffsets to avoid this problem，but still read PCFile

this problem lead to your testcase TestJitPanicRecoveryStackTrace failed

pkujhd commented:
> I'd be curious how you plan to address the incorrect cuFiles and cuOffsets after the pclntab separation causing broken stack traces

I guess what you mean is the error of PCFile. 
There is an error in the previous version. 

In the Golang<1.16 version, It leads to the function which isn't in the first .obj file has not correct offset of the fileno.
In version>=1.16, the funcfile function will be panic

In the version of golang<1.16 PCFile will be rewritten by linker,
In the version of>=1.16, linker uses cuOffset to deal it

I'm going to use a similar scheme to avoid mistakes

Commit:
Bump JIT go.mod after hard fork

Commit:
Bake in a bit more of reflect

Commit:
Revert fed0a9e753219949a4bb1e6b5a6e5a3a99dffc7e as it was simply masking an upstream bug in symRef resolution when reading an archive.

Not all symRefs will have their name included in an archive's refNames, especially if they're `goobj.AuxGotype`s which point to a type from another package, so we need to keep track of any unresolved symRefs and try to resolve them all after we've loaded all archive files.

A typical bug case was where a package declared a global `sync.Map` (or any global whose type is from another package). These globals would be GC roots without any gcmasks/gcprogs since the type information for the global was not resolved correctly.

This commit also adds a JIT test case which probes this behaviour, and panics due to GC memory corruption before this change

eh-steve committed:
Merge pull request #2 from eh-steve/bugfix-gcdata-revert-fed0a9e

Bugfix gcdata revert fed0a9e

eh-steve committed:
Type deduplication safety improvements (#3)

* Improve firstmodule type registration by generating symbol names from type names correctly - this allows us to full remove the second pass of type deduplication using type hashes of typelinks.

* Add test demonstrating risk (fix still TODO)

* Earlier attempts at single pass relocation were misguided...

* Fix bug with checking for firstmodule type variants

* Force rebuild of non-stdlib types only

* Just use go list directly instead of golang.org/x/tools/go/packages

* Use the same version of go toolchain for finding stdlibs as for building JIT code

* Consolidate different ways of registering types

* Further type registration bugfixes

* First attempt at making symbols store their package if possible, and JIT deps builder build according to this package instead of previous fuzzy matching. Also only force build of types reachable from the main package (first package in loader)

* Add specialised unloaders for safely removing traces of JIT code in host package global caches

* Fix wrong commit in go.mod

* Allow elements of arrays to be converted

---------

Co-authored-by: Anonymous <anonymous@github.com>

invalid-email-address committed:
WIP - assuming gc is patched to emit types for exported functions, build a map[string]interface{} of package exports directly

invalid-email-address committed:
Add PatchGC() function and simplify JIT builds (no copy)

invalid-email-address committed:
Always use heap strings (remove config) - required bugfix of epilogue size maxExtraInstructionBytesPCRELxJMP

Commit:
Add bakeInPlatform() for OS/arch specific things

invalid-email-address committed:
Bump jit go.mod

invalid-email-address committed:
Bake in a few more types

invalid-email-address committed:
Make patch allow moves across unix filesystem boundaries

invalid-email-address committed:
Store ·f function containers in CodeModule.Syms for use in inter-module type conversion

invalid-email-address committed:
Add some GCs to TestJitComplexFunctions in case a JIT net poll is still running in the background since the dialler's deadline ctx would have skipped over it immediately even if it was still polling

invalid-email-address committed:
Add CLI tool for gc patching (easier to run via sudo)

invalid-email-address committed:
Try removing replaces

invalid-email-address committed:
Update error message

invalid-email-address committed:
Update README.md

invalid-email-address committed:
Delete goloader examples

invalid-email-address committed:
Delete old compatibility files for now unsupported Go versions

invalid-email-address committed:
Bump go.mod/go.sum

invalid-email-address committed:
Improve error messages + make sure type deduplication relocation errors cause a failure

eh-steve committed:
Update go.yml

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Fucking windows...

invalid-email-address committed:
Bump go.mod again

invalid-email-address committed:
Bump go.mod again

invalid-email-address committed:
Force building of packages for named (resolved) symbols but whose pkgIdx is None by adding dummy R_KEEP relocation

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Update http tests

invalid-email-address committed:
Test for Issue #74

invalid-email-address committed:
Fix another type registration bug

invalid-email-address committed:
Simplify registerType(), and use the exact same recursive approach for buildModuleTypeHash()

invalid-email-address committed:
Bugfix inlinetree - attempt to use inlined func's funcID, otherwise assume normal

invalid-email-address committed:
Properly add both types of ABI wrapper funcs with the correct abi0 or abiinternal suffix to avoid accidentally relocating into a wrapper

invalid-email-address committed:
itab.go - rebase ifn/tfn offsets according to whether the type is in first or JIT module

invalid-email-address committed:
Always pointer align data as even on amd64, certain bulk memmove operations expect pointer alignment

invalid-email-address committed:
Expose linker.Autolib()

invalid-email-address committed:
Bugfix addSymbolMap() where duplicate functions would always point firstmodule prefixed version at new version

invalid-email-address committed:
Rudimentary reachability analysis/deadcode elimination.

+ traverse firstmodule itabs to force type/package rebuild if new module has R_USEIFACEMETHOD relocs, and this method is unreachable in original firstmodule

invalid-email-address committed:
JIT - remove forbidden packages (this was previously a problem because of incorrect ABI0/ABIInternal relocation, now it's safe to rebuild any package), and therefore also remove baked in functions/types.

+ restructure test config

+ add tests for big packages (k8s)

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Skip k8s tests according to go version

invalid-email-address committed:
Placeholder func for PE relocs

invalid-email-address committed:
Fix libc_error symbol lookup

invalid-email-address committed:
Add CGO_ENABLED={0,1} to workflow matrix

invalid-email-address committed:
darwin/arm64 requires CGO

invalid-email-address committed:
Share build env

invalid-email-address committed:
Bugfix use of copy2Slice instead of putAddress

eh-steve committed:
Update jit/jit.go

invalid-email-address committed:
Add -dynlink by default and add support for R_GOTPCREL and R_TLS_IE (arm64 still TODO)

Add test for issue #78

invalid-email-address committed:
Make issue 78 test demonstrate the problem with PCREL MOVs implemented how they were

invalid-email-address committed:
Multiple fixes:
 - rewrite PCREL reloc assembly to be correct
 - JIT uses -dynlink by default (TODO - implement GOTPCREL relocs for arm64)
 - make issue #78 test useful in demonstrating the desired behaviour
 - define x86 epilogue sizes in terms of asm []byte lengths

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Express arm epilogue sizes in terms of asm sizes and add reloctype R_ARM64_GOTPCREL (not implemented yet)

invalid-email-address committed:
Add another test case for issue #77

invalid-email-address committed:
Fix R_TLS_IE when CGo is disabled

invalid-email-address committed:
Don't escape go.shape pkgPath during type registration

invalid-email-address committed:
Actually implement R_ARM64_GOTPCREL relocs, and make the use of -dynlink gcflag configurable in jit package.

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
More PCREL arithmetic mistakes!

invalid-email-address committed:
Further bugfix PCREL CMPL code (needed indirect), and MOV code (incomplete dst register bits). Fill epilogues and alignment padding with NOP instructions for STEXT to simplify disassembly

invalid-email-address committed:
Don't attempt to register missing types with no package as CGo symbols

invalid-email-address committed:
Further tweaks to CMPL PCREL reloc asm

invalid-email-address committed:
Add concurrency limit of GOMAXPROCS to jit dependency builds

invalid-email-address committed:
Add ForceTestRelocationEpilogues linker option

invalid-email-address committed:
Experiment - try using handwritten asm wrappers to call libdl on darwin when CGo is disabled

invalid-email-address committed:
Add CGo-less darwin/arm64 support

invalid-email-address committed:
Bump go.mod

eh-steve committed:
Update README.md (optimistically!)

invalid-email-address committed:
bugfix darwin/arm64 go1.18 inter-module closure conversion

invalid-email-address committed:
Deterministic symbol resolution order

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
When testing jit.BuildGoText, cd into the testdata module to avoid polluting the jit package's go.mod/go.sum

invalid-email-address committed:
If a symbol from a package is still missing after loading that package, take a guess that it's linknamed and implemented in runtime

invalid-email-address committed:
Finish off minimal PE relocs implementation

invalid-email-address committed:
Skip k8s build on windows due to golang/x/sys/windows init() being unable to find kernel32.dll (possibly a problem with that package). Also ignore runtime.tls_g symbol found on windows

invalid-email-address committed:
Skip PE reloc support for windows on go1.18 + skip CGo test

eh-steve committed:
Test for issue #75 (#8)

* Test for issue #75

* Make test actually profile

---------

Co-authored-by: Anonymous <anonymous@github.com>

invalid-email-address committed:
Skip another test on windows

invalid-email-address committed:
If a symbol we suspect of being linknamed and implemented in runtime was not found in runtime, check whether it's a runtime/internal/* type and use that as it's package (goobj symrefs don't seem to give the correct package for these types?)

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Another type registration bugfix

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Simplify PCREL CGo CALLs

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Simplify PCREL CGo CALLs

invalid-email-address committed:
Use reflect.(*rtype).Elem during type registration

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
WIP - support Go 1.21 (RC3)

invalid-email-address committed:
Handle errors during regSymbol(), and register RODATA too

invalid-email-address committed:
Fix mistake in symbol size calculation

invalid-email-address committed:
Add protobuf test

invalid-email-address committed:
Force static data to be marked as reachable to ensure packages containing their types are built

invalid-email-address committed:
Patch firstmodule type's method types if marked unreachable (-1), and hijack firstmoduledata.typemap to allow resolveTypeOff to find JIT types using a firstmodule base address

invalid-email-address committed:
Fix ordering of inittasks via Autolib() order (not R_INITORDER), also fix bug where inittask package names containing a dot were not properly escaped

invalid-email-address committed:
Add Issue55 version of TestPatchMultipleModuleItabs

invalid-email-address committed:
Fix protobufunload and mistake in DataAddr()

invalid-email-address committed:
Fix Macho relocs + WIP support for more elf and macho arm64 reloc types

invalid-email-address committed:
Fix PE relocs (to support .bss sections) + add more substantial Cgo test which involves mutating C initialised data and C calling Go

invalid-email-address committed:
Make 2 versions x86 asm on darwin/amd64, since on 1.21, the toolchain wraps handwritten asm with a PUSH/POP BP to create frames for you

invalid-email-address committed:
Skip advanced CGo test in CI

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Bump patchgc

invalid-email-address committed:
Add test for issue 87

invalid-email-address committed:
Fix patchgc for cross-device copies on Windows

invalid-email-address committed:
Bump patchgc

invalid-email-address committed:
Update CALL relocs - try JMP (to epilogue) then CALL (target func) then JMP (back from epilogue) instead of CALL (into epilogue) then JMP (into target func), since if a traceback hits while we're in the epilogue, the funcForPC for the current PC will not be aligned with the link register (regardless whether the LR is virtual or real), and it will appear that the callee and caller are the same `_func`

invalid-email-address committed:
Fix darwin/macho self patch when os.Args[0] is not an absolute or relative path (e.g. when resolved via $PATH)

invalid-email-address committed:
fix #91, when pcdata is empty, so pcdata offset is zero

invalid-email-address committed:
fix #93, fix fill findfuncbucket error

invalid-email-address committed:
Improve type registration and behaviour under 1.21

invalid-email-address committed:
Make go toolchain commands output to stderr/stdout if DebugLog = true

invalid-email-address committed:
Try go test -a -c for rebuilding stdlibs on Windows?

invalid-email-address committed:
Adjust readFuncData to add a 0 for empty string funcdata

invalid-email-address committed:
Run patchgc with -a to force rebuild of stdlib on Windows

invalid-email-address committed:
Add test for type aliases to pointer types

invalid-email-address committed:
Don't write GoList output to stdout

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
Don't stream IO in GoList

invalid-email-address committed:
Improve error in patchgc

invalid-email-address committed:
Change API of ConvertTypesAcrossModules to make it clearer that old value gets converted to new type

invalid-email-address committed:
Bump go.mod

invalid-email-address committed:
(Untested) Go 1.22 support

pkujhd commented:
The current master version has basically completed the support for cgo.
For the tls problem, it should only appears in the -dynlink compilation mode of Linux. 
Currently, for linux/amd64,  the goloaderbuilder does not generate position-independce-code,  generates position-independce-code on windows/mac platform.
For the duplicate type problem, the current version defaults to using the type in the original running environment, and adds fake itab through interface detection to avoid panic with unreachable code. This solution should be safer than directly modifying the read-only memory of the runtime in your fork.
In addition, on the x86 platform, in order to reduce the problems caused by modifying the machine code, golang/x/arch/x86/x86asm is used to decode machine code and mark the instruction boundary.

So I will close this pull reqeust.

Thank you for your contribute.

There are still some remaining issues. There is no good solution to the initialization problem of the current high golang version(>=1.21). The golang linker uses the compiler to assist in trimming the initialization function to versions >= 1.21.
